### PR TITLE
Add a `/lookup` endpoint

### LINKED ISSUE #572: Add an API endpoint for identifier-based lookups
When debugging a query (e.g. #284), it can be difficult to figure out whether there is an error in CAM-KP-API itself or in the underlying database. It would be helpful to add an API endpoint for identifier-based lookups. I'm imagining a simple RESTful API such as:

* `/lookup/?subject=http://identifiers.org/ncbigene/1956&predicate=https://w3id.org/biolink/vocab/decreases_amount_or_activity_of&object_category=https://w3id.org/biolink/vocab/ChemicalSubstance`: return everything we know about NCBIGene:1956 biolink:decreases_amount_or_activity_of of any biolink:ChemicalSubstance.
* `/lookup/NCBIGene:1956`: return everything we know about NCBIGene:1956
* `/lookup/NCBIGene:1956/biolink:decreases_amount_or_activity_of`: return everything we know NCBIGene:1956 decreases amount or activity of.

Since this API will side-step the TRAPI processing code, we can use it to determine exactly what the backend triplestore knows about the queried identifier in a way that is easier to use than querying the triplestore directly.

### PR DESCRIPTION
**Pull request by gaurav**
This PR adds a `/lookup` endpoint that will allow us to find out what CAM-KP-API knows about provided identifiers. This has now been deployed (as v0.3-pre1) to https://cam-kp-api-dev.renci.org/1.3.0/docs/index.html?url=docs.yaml#/default/getLookup (using the old development SPARQL, not the new development SPARQL, which has issues -- see https://github.com/ExposuresProvider/cam-pipeline/issues/80). It includes a test that has a number of identifiers to try out; this includes one identifier that can't be normalized (but are in CAM-KP).

Closes #572.

### CODE REVIEW

File: src/main/scala/org/renci/cam/LookupService.scala [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+      subjectIRIs = {
+        implicit val iriKeyDecoder: KeyDecoder[IRI] = Implicits.iriKeyDecoder(biolinkData.prefixes)
+
+        qualifiedIds.map(qualifiedId => iriKeyDecoder(qualifiedId).get)
+      }
+
+      relationsQuery =
+        sparql"""PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+               PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
```
Comment (balhoff) [inline src/main/scala/org/renci/cam/LookupService.scala:157]: Don't use prefixes in sparql interpolator queries. Just embed the constant IRI object where needed.

File: src/main/scala/org/renci/cam/LookupService.scala [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+
+               SELECT DISTINCT ?subj ?subjLabel ?p ?pLabel ?obj ?objLabel ?g {
+                  ?s <http://www.openrdf.org/schema/sesame#directType> ?subj .
+                  ?subj <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?subjClass .
```
Comment (balhoff) [inline src/main/scala/org/renci/cam/LookupService.scala:161]: Embed constant IRIs for predicates.

Comment (balhoff) [reply src/main/scala/org/renci/cam/LookupService.scala:161]: (and other places in the query)

File: src/main/scala/org/renci/cam/LookupService.scala [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    o: String,
+    g: String
+  )
+
+  /** A helper method for converting a QuerySolution with ?s, ?p, ?o, ?g into a ResultTriple. */
+  def fromQuerySolution(result: QuerySolution): ResultTriple = ResultTriple(
+    result.get("s").toString,
+    result.get("p").toString,
+    result.get("o").toString,
+    result.get("g").toString
```
Comment (balhoff) [inline src/main/scala/org/renci/cam/LookupService.scala:57]: Are these meant to be just any kind of RDF node? Not necessarily IRI or literal or whatever?

File: src/main/scala/org/renci/cam/LookupService.scala [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+               }"""
+      relationsResults <- SPARQLQueryExecutor.runSelectQuery(relationsQuery.toQuery)
+      preds = relationsResults.map(_.getResource("p").getURI).map(IRI(_))
+      biolinkRelationMap <- QueryService.mapRelationsToLabelAndBiolink(preds.toSet)
+
+      // We want to group these by object, so we don't return a gazillion predicates for each result.
+      objectMap = relationsResults.groupBy(_.getResource("obj").getURI)
+
+      relations = objectMap map { case (obj, results) =>
+        val predResults = results map { res =>
```
Comment (balhoff) [inline src/main/scala/org/renci/cam/LookupService.scala:183]: We can discuss, but I only really use infix syntax for DSL situations.

File: src/main/scala/org/renci/cam/LookupService.scala [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+    * @param conflate
+    *   What to set in the `conflate` param to NodeNorm. Currently, may be `true` or `false`.
+    */
+  def getQualifiedIdsFromNodeNorm(queryId: String,
+                                  nodeNormURL: String,
+                                  conflate: String): ZIO[Blocking with Has[BiolinkData], Exception, List[LabeledIRI]] =
+    for {
+      biolinkData <- biolinkData
+      nnUri <- ZIO.fromEither(Uri.fromString(nodeNormURL))
+      uri = nnUri.+?("curie", queryId).+?("conflate", conflate).toString()
```
Comment (balhoff) [inline src/main/scala/org/renci/cam/LookupService.scala:269]: What is `+?`?

File: src/main/scala/org/renci/cam/LookupService.scala [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+        implicit val iriEncoder: Encoder[IRI] = Implicits.iriEncoder(biolinkData.prefixes)
+
+        implicit val iriKeyEncoder: KeyEncoder[IRI] = Implicits.iriKeyEncoder(biolinkData.prefixes)
+        implicit val iriKeyDecoder: KeyDecoder[IRI] = Implicits.iriKeyDecoder(biolinkData.prefixes)
+
+        implicit val biolinkClassEncoder: Encoder[BiolinkClass] = Implicits.biolinkClassEncoder
+        implicit val biolinkClassDecoder: Decoder[BiolinkClass] = Implicits.biolinkClassDecoder(biolinkData.classes)
+
+        implicit val biolinkPredicateEncoder: Encoder[BiolinkPredicate] = Implicits.biolinkPredicateEncoder(biolinkData.prefixes)
+        implicit val biolinkPredicateDecoder: Decoder[List[BiolinkPredicate]] =
```
Comment (balhoff) [inline src/main/scala/org/renci/cam/LookupService.scala:290]: Just a comment—we do this in a lot of places. We should make a class which we can construct from biolink data, whose objects provide implicit fields that can just be imported, like:

```scala
val de = DecoderEncoder(biolinkData)
import de._
```

File: src/main/scala/org/renci/cam/LookupService.scala [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+import sttp.tapir.Endpoint
+import sttp.tapir.generic.auto._
+import sttp.tapir.server.http4s.ztapir.ZHttp4sServerInterpreter
+import sttp.tapir.ztapir.{endpoint, query}
+import zio.blocking.{effectBlockingIO, Blocking}
+import zio.config.ZConfig
+import zio.{Has, RIO, URIO, ZIO}
+
+import scala.io.Source
+import scala.language.implicitConversions
```
Comment (balhoff) [inline src/main/scala/org/renci/cam/LookupService.scala:24]: What's this for?

**Review by balhoff — CHANGES_REQUESTED** [review]

File: src/main/scala/org/renci/cam/LookupService.scala (hunk 1 above)
Comment (gaurav) [reply src/main/scala/org/renci/cam/LookupService.scala]: Fixed in a664bdc.

File: src/main/scala/org/renci/cam/LookupService.scala (hunk 2 above)
Comment (gaurav) [reply src/main/scala/org/renci/cam/LookupService.scala]: Fixed in c90c3d8.

File: src/main/scala/org/renci/cam/LookupService.scala (hunk 3 above)
Comment (gaurav) [reply src/main/scala/org/renci/cam/LookupService.scala:57]: Yup! I've added a comment to explain this in cba4556.

File: src/main/scala/org/renci/cam/LookupService.scala (hunk 4 above)
Comment (gaurav) [reply src/main/scala/org/renci/cam/LookupService.scala]: In turns out this wasn't entirely my fault -- this is how scalafmt formats this with our settings. I modified .scalafmt.conf (3249bcfd7b1e0) to add ApplyInfix and re-ran scalafmt, which partially fixed this to `objectMap.map { case (obj, results) =>` (a1826d981b0). Do you think that's good enough, or should we try to figure out how to modify .scalafmt.conf to add a curly bracket there?

File: src/main/scala/org/renci/cam/LookupService.scala (hunk 5 above)
Comment (gaurav) [reply src/main/scala/org/renci/cam/LookupService.scala]: Replaced with the easier to read `.withQueryParam()` in 250c718.

File: src/main/scala/org/renci/cam/LookupService.scala (hunk 6 above)
Comment (gaurav) [reply src/main/scala/org/renci/cam/LookupService.scala]: Merged master in 408d492e9a54bfc2 and replaced this in 83bf5e4.

File: src/main/scala/org/renci/cam/LookupService.scala (hunk 4 above)
Comment (balhoff) [reply src/main/scala/org/renci/cam/LookupService.scala]: That looks good!

**Review by balhoff — APPROVED** [review]

**Pull request merged**